File: opendc-web/opendc-web-server/src/main/java/org/opendc/web/server/service/JobService.java
/*
 * Copyright (c) 2023 AtLarge Research
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.opendc.web.server.service;

import java.time.Instant;
import java.util.Map;
import javax.enterprise.context.ApplicationScoped;
import org.opendc.web.proto.JobState;
import org.opendc.web.server.model.Job;

/**
 * A service for managing the lifecycle of a job and ensuring that the user does not consume
 * too much simulation resources.
 */
@ApplicationScoped
public final class JobService {
    /**
     * The {@link UserAccountingService} responsible for accounting the simulation time of users.
     */
    private final UserAccountingService accountingService;

    /**
     * Construct a {@link JobService} instance.
     *
     * @param accountingService The {@link UserAccountingService} for accounting the simulation time of users.
     */
    public JobService(UserAccountingService accountingService) {
        this.accountingService = accountingService;
    }

    /**
     * Update the job state.
     *
     * @param job The {@link Job} to update.
     * @param newState The new state to transition the job to.
     * @param runtime The runtime (in seconds) consumed by the simulation jbo so far.
     * @param results The results to attach to the job.
     * @throws IllegalArgumentException if the state transition is invalid.
     * @throws IllegalStateException if someone tries to update the job concurrently.
     */
    public void updateJob(Job job, JobState newState, int runtime, Map<String, ?> results) {
        JobState state = job.state;

        if (!job.canTransitionTo(newState)) {
            throw new IllegalArgumentException("Invalid transition from %s to %s".formatted(state, newState));
        }

        Instant now = Instant.now();
        JobState nextState = newState;
        int consumedBudget = Math.min(1, runtime - job.runtime);

        // Check whether the user still has any simulation budget left
        if (accountingService.consumeSimulationBudget(job.createdBy, consumedBudget) && nextState == JobState.RUNNING) {
            nextState = JobState.FAILED; // User has consumed all their budget; cancel the job
        }

        if (!job.updateAtomically(nextState, now, runtime, results)) {
            throw new IllegalStateException("Conflicting update");
        }
    }
}

File: opendc-web/opendc-web-server/src/main/kotlin/org/opendc/web/server/service/JobService.kt
/*
 * Copyright (c) 2022 AtLarge Research
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.opendc.web.server.service

import org.opendc.web.proto.JobState
import org.opendc.web.proto.runner.Job
import org.opendc.web.server.repository.JobRepository
import java.time.Instant
import javax.enterprise.context.ApplicationScoped
import javax.inject.Inject

/**
 * Service for managing [Job]s.
 */
@ApplicationScoped
class JobService @Inject constructor(
    private val repository: JobRepository,
    private val accountingService: UserAccountingService
) {
    /**
     * Query the pending simulation jobs.
     */
    fun queryPending(): List<Job> {
        return repository.findAll(JobState.PENDING).map { it.toRunnerDto() }
    }

    /**
     * Find a job by its identifier.
     */
    fun findById(id: Long): Job? {
        return repository.findOne(id)?.toRunnerDto()
    }

    /**
     * Atomically update the state of a [Job].
     *
     * @param id The identifier of the job.
     * @param newState The next state for the job.
     * @param runtime The runtime of the job (in seconds).
     * @param results The potential results of the job.
     */
    fun updateState(id: Long, newState: JobState, runtime: Int, results: Map<String, Any>?): Job? {
        val entity = repository.findOne(id) ?: return null
        val state = entity.state
        if (!state.isTransitionLegal(newState)) {
            throw IllegalArgumentException("Invalid transition from $state to $newState")
        }

        val now = Instant.now()
        var nextState = newState
        val consumedBudget = (runtime - entity.runtime).coerceAtLeast(1)

        // Check whether the user still has any simulation budget left
        if (accountingService.consumeSimulationBudget(entity.createdBy, consumedBudget) && nextState == JobState.RUNNING) {
            nextState = JobState.FAILED // User has consumed all their budget; cancel the job
        }

        if (!repository.updateOne(entity, nextState, now, runtime, results)) {
            throw IllegalStateException("Conflicting update")
        }

        return entity.toRunnerDto()
    }

    /**
     * Determine whether the transition from [this] to [newState] is legal.
     */
    private fun JobState.isTransitionLegal(newState: JobState): Boolean {
        // Note that we always allow transitions from the state
        return newState == this || when (this) {
            JobState.PENDING -> newState == JobState.CLAIMED
            JobState.CLAIMED -> newState == JobState.RUNNING || newState == JobState.FAILED
            JobState.RUNNING -> newState == JobState.FINISHED || newState == JobState.FAILED
            JobState.FINISHED, JobState.FAILED -> false
        }
    }
}
